article: How to migrate from IDF Bootloader to MCUboot using ESP Self-Reflasher (#363)
espressif/developer-portal: content/blog/2024/12/how-to-migrate-from-idf-bootloader-to-mcuboot-using-esp-self-reflasher/index.md

---

Migration in Embedded Systems is always a significant subject and it may not only feature on an ongoing production decision, but also as a factor that can contribute to design choices at project time. Hardware or firmware wise, any change of that kind is risky, so it requires good planning, controled testing and careful execution.

On the firmware area, there are a variety of reasons that may motivate choosing a flexible platform for a project or actually migrating from a production running RTOS (Real Time Operating System) to another one, they may range from technical to strategic purposes.

Relatively, changing a RTOS may not be as straight forward as a firmware update. One of the concerns that needs to be addressed beforehand is the bootloader compatibility. Besides initializing the system, the bootloader may be responsible for handling safe and secure firmware updates, however usually it cannot update itself or it is not "updatable" by design or due device restriction features, e.g. flash bootloader region protection. Also in most of the scenarios, the device is not physically accessible, thus a direct serial flashing is not practical.

> This tutorial will demonstrate, observed the constraints, how ESP Self-Reflasher component can be used to migrate an ESP32 embedded with an IDF-based application (OTA capable) + IDF bootloader to MCUboot compatible application + MCUboot bootloader. The ESP Self-Reflasher component is used on an middle step application for enabling the device "reflash" its own firmware.

Reviewer: ```suggestion
This tutorial will demonstrate, how ESP Self-Reflasher component can be used to migrate an ESP32 embedded with an IDF-based application (OTA capable) + IDF bootloader to MCUboot compatible application + MCUboot bootloader. The ESP Self-Reflasher component is used on an middle step application for enabling the device "reflash" its own firmware.
```

> This tutorial will demonstrate, observed the constraints, how ESP Self-Reflasher component can be used to migrate an ESP32 embedded with an IDF-based application (OTA capable) + IDF bootloader to MCUboot compatible application + MCUboot bootloader. The ESP Self-Reflasher component is used on an middle step application for enabling the device "reflash" its own firmware.

Reviewer: ```suggestion
This tutorial will demonstrate how ESP Self-Reflasher component can be used to migrate an ESP32 embedded with an IDF-based application (OTA capable) + IDF bootloader to MCUboot compatible application + MCUboot bootloader. The ESP Self-Reflasher component is used on an middle step application for enabling the device to "reflash" its own firmware.
```

The [ESP Self-Reflasher](https://components.espressif.com/components/espressif/esp-self-reflasher/versions/1.0.0) is an ESP-IDF component that allows the device to reflash itself with new firmware. This is achieved through one of the following ways of operation: downloading the reflashing image or embedding the reflashing image. The component has examples for each that will be used on this guide.

## Environment pre requisites

Besides ESP-IDF ([Get started guide here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/index.html)), the following are required:

- ESP Self-Reflasher: clone it to `<IDF_DIR>/components` with Git:

```bash
cd <IDF_DIR>/components
git clone https://github.com/espressif/esp-self-reflasher.git
```

- NuttX: in order to build the application that will be migrated to, set the NuttX workspace. [Get started guide here](https://developer.espressif.com/blog/nuttx-getting-started/).

- MCUboot: in the case of building the bootloader standalone, follow the instructions [here](https://docs.mcuboot.com/readme-espressif.html).

## Step-by-step Guide

### Building the reflashing images (MCUboot and NuttX application)

> Assuming that the NuttX workspace is set, prepare the NuttX application that will be the final target of this process:
> 
> The application that will be used is the NSH configured for MCUboot compatibility:
> 

Reviewer: ```suggestion
With the NuttX workspace ready, prepare the NSH configuration with MCUboot compatibility, which is available under the `mcuboot_nsh` defconfig. This is the build that will be used on the reflashing process.

```

```bash
cd <NUTTX_DIR>
./tools/configure.sh esp32-devkitc:mcuboot_nsh
```

> However, it is possible to manually configure the example for building it with MCUboot compatibility. Enter the menuconfig:

Reviewer: So this section is only if the user is not using the `mcuboot_nsh` right?

Author: Yes, I'll emphasize that on the text formatting.

| REGION                           | OFFSET    | SIZE     |
| -------------------------------- | --------- | -------- |
| Bootloader                       | 0x001000  | 0x00F000 |
| Application primary slot         | 0x010000  | 0x100000 |
| Application secondary slot       | 0x110000  | 0x100000 |
| Scratch                          | 0x210000  | 0x040000 |

---
If MCUboot is being build separately from NuttX build, be sure that the configuration for flash organization matches on both.

---

Save and exit the menuconfig.

Build the MCUboot bootloader:

```bash
make -j bootloader
```

Now build the NuttX application:

```bash
make ESPTOOL_BINDIR=./ -s -j
```

### Building the Reflashing application

#### Example 1: Download the Target Reflashing Image

ESP Self-Reflasher component provides the example `boot_swap_download_example`, this can be used as start point for the Reflashing application.

```bash
cd <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_download_example/
```

> Firstly, copy the final target images to the directory that will be served for HTTP download.

Reviewer: ```suggestion
First, copy the NuttX and MCUBoot binaries to the directory that will be served for HTTP download.
```

```bash
cp <NUTTX_DIR>/nuttx.bin <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_download_example/example_bin_dir/app_upd.bin
cp <NUTTX_DIR>/mcuboot-esp32.bin <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_download_example/example_bin_dir/mcuboot-esp32.bin
```

Configure the example:

```bash
idf.py menuconfig
```

Enter menu `Example Configuration -> Bootloader reflash configuration` and set the `Bootloader bin url endpoint` with the server IP, port and the bootloader binary name (in this example, "mcuboot-esp32.bin"):

```
http://<SERVER_IP>:<PORT>/mcuboot-esp32.bin
```

Also configure the `Bootloader region address` (the Bootloader offset mentioned above) and the `Bootloader region size`.

Then configure the similar parameters on the `Example Configuration -> Application reflash configuration` menu, set the `Application upgrade url endpoint`:

```

> http://<SERVER_IP>:<PORT>/app_upd.bin

Reviewer: What should be the IP and port? Is it the user computer IP?

Author: Changed it to HOST_IP

---
title: "How to migrate from IDF Bootloader to MCUboot using ESP Self-Reflasher"
date: 2024-12-16
showAuthor: false
authors:
  - "almir-okato"
tags: ["Component", "ESP-IDF", "Bootloader", "MCUboot", "ESP Self Reflasher", "NuttX"]
---

Migration in Embedded Systems is always a significant subject and it may not only feature on an ongoing production decision, but also as a factor that can contribute to design choices at project time. Hardware or firmware wise, any change of that kind is risky, so it requires good planning, controled testing and careful execution.

On the firmware area, there are a variety of reasons that may motivate choosing a flexible platform for a project or actually migrating from a production running RTOS (Real Time Operating System) to another one, they may range from technical to strategic purposes.

> Relatively, changing a RTOS may not be as straight forward as a firmware update. One of the concerns that needs to be addressed beforehand is the bootloader compatibility. Besides initializing the system, the bootloader may be responsible for handling safe and secure firmware updates, however usually it cannot update itself or it is not "updatable" by design or due device restriction features, e.g. flash bootloader region protection. Also in most of the scenarios, the device is not physically accessible, thus a direct serial flashing is not practical.

Reviewer 2: `changing an RTOS`

Reviewer 2: "However, it typically cannot update itself, either by design or due to device restrictions, such as flash bootloader region protection.""

> This tutorial will demonstrate, how ESP Self-Reflasher component can be used to migrate an ESP32 embedded with an IDF-based application (OTA capable) + IDF bootloader to MCUboot compatible application + MCUboot bootloader. The ESP Self-Reflasher component is used on an middle step application for enabling the device to "reflash" its own firmware.

Reviewer 2: "This tutorial demonstrates how ESP Self-Reflasher.."

---
**Note**
The MCUboot compatible application used in this guide is a NuttX RTOS application, but could be other like a Zephyr RTOS application, for instance.

> 

Reviewer 2: Is this extra line needed?

Author: Yes, I tried removing the line and the format broke.

---

The overall process described in this guide can be illustrated as:

```mermaid
flowchart LR
    A[IDF bootloader + IDF application with OTA capabilities] -->|OTA process| B[Reflashing application with ESP Self-Reflasher component]
    B -->|Reflashing process| C(MCUboot bootloader + NuttX application)
```

## ESP Self Reflasher component

> The [ESP Self-Reflasher](https://components.espressif.com/components/espressif/esp-self-reflasher/versions/1.0.0) is an ESP-IDF component that allows the device to reflash itself with new firmware. This is achieved through one of the following ways of operation: downloading the reflashing image or embedding the reflashing image. The component has examples for each that will be used on this guide.

Reviewer 2: I would re-write this part to something similar to:
"..to be reflashed with new firmware."

## Environment pre requisites

> Besides ESP-IDF ([Get started guide here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/index.html)), the following are required:

Reviewer 2: "Besides ESP-IDF framework"

Besides initializing the system, the bootloader may be responsible for handling safe and secure firmware updates. However, it typically cannot update itself, either by design or due to device restrictions, such as flash bootloader region protection. Also in most of the scenarios, the device is not physically accessible, thus a direct serial flashing is not practical.

This tutorial demonstrates how ESP Self-Reflasher component can be used to migrate an IDF-bootloader-based application (OTA capable) to a MCUboot-based-bootloader application. The ESP Self-Reflasher component is used on an middle step application for enabling the device to "reflash" its own firmware.

---
**Note**
The MCUboot compatible application used in this guide is a NuttX RTOS application, but could be other like a Zephyr RTOS application, for instance.

---

The overall process described in this guide can be illustrated as:

```mermaid
flowchart LR
    A[IDF bootloader + IDF application with OTA capabilities] -->|OTA process| B[Reflashing application with ESP Self-Reflasher component]
    B -->|Reflashing process| C(MCUboot bootloader + NuttX application)
```

## ESP Self Reflasher component

> The [ESP Self-Reflasher](https://components.espressif.com/components/espressif/esp-self-reflasher/versions/1.0.0) is an ESP-IDF component that allows the device to be "reflashed" with new firmware. This is achieved through one of the following ways of operation: downloading the reflashing image or embedding the reflashing image. The component has examples for each that will be used on this guide.

Author: For some reason the component page didn't show the README file, I guess it doesn't update information after version change on `idf_component.yml`. I've updated it again, now its properly shown.

## Environment pre requisites

Besides ESP-IDF ([Get started guide here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/index.html)), the following are required:

- ESP Self-Reflasher: clone it to `<IDF_DIR>/components` with Git:

```bash
cd <IDF_DIR>/components
git clone https://github.com/espressif/esp-self-reflasher.git
```

- NuttX: in order to build the application that will be migrated to, set the NuttX workspace. [Get started guide here](https://developer.espressif.com/blog/nuttx-getting-started/).

- MCUboot: in the case of building the bootloader standalone, follow the instructions [here](https://docs.mcuboot.com/readme-espressif.html).

## Step-by-step Guide

> ### Building the reflashing images (MCUboot and NuttX application)

Reviewer 3: The procedures are commonly titled with verbs in their infinitive form to make the title focus on action almost like a command:

- Build the reflash images (MCUboot and NuttX application)
- Build the reflashing application
- Build and flash the IDF application

With the NuttX workspace ready, prepare the NSH (NuttShell) configuration with MCUboot compatibility, which is available under the `mcuboot_nsh` defconfig. This is the build that will be used as the *final target* of the reflashing process.

```bash
cd <NUTTX_DIR>
./tools/configure.sh esp32-devkitc:mcuboot_nsh
```

> ---
> **Note**
> 
> It is also possible to manually configure other NuttX applications for building it with MCUboot compatibility. Enter the menuconfig:
> 
> ```bash
> make menuconfig
> ```
> 
> Enter the menu `System Type -> Bootloader and Image Configuration` and enable the `Enable MCUboot-bootable format` option.
> 
> ---

Reviewer 3: Why are you placing this information in a note? It seems to be as important and as relevant as any other information.

Also suggest optimizing the formatting:

````suggestion
You can also manually configure other NuttX applications to be built with MCUboot compatibility:

- In your NuttX application, enter the menuconfig:
  ```bash
  make menuconfig
  ```
- In the menu `System Type -> Bootloader and Image Configuration`, enable the option `Enable MCUboot-bootable format`
````

> Ensure that the organization of image slots in flash is properly configured. It can be checked on the menu `System Type -> SPI Flash Configuration` from menuconfig. For this example using ESP32, the following is considered (note that the Bootloader offset may differ among other Espressif chip families):

Reviewer 3: I assume `organization of image slots` means `partition table`. If yes, please use more established terminology.

Not clear what the verb _considered_ in `the following is considered` means. Should it be _used_ or _applied_?

Suggested update (please feel free to further edit it):

```suggestion
Ensure that the flash partition table is properly configured in menuconfig -> `System Type -> SPI Flash Configuration`. For this example using the ESP32, the following is partition table is used (note that the Bootloader offset may differ among other Espressif chip families):
```

Author: MCUboot doesn't use the "partition table" term, that's why I avoided it. 
But now I changed it to "flash map" now, as can be found in MCUboot documentation.

| REGION                           | OFFSET    | SIZE     |
| -------------------------------- | --------- | -------- |
| Bootloader                       | 0x001000  | 0x00F000 |
| Application primary slot         | 0x010000  | 0x100000 |
| Application secondary slot       | 0x110000  | 0x100000 |
| Scratch                          | 0x210000  | 0x040000 |

> ---
> **Note**
> 
> If MCUboot is being build separately from NuttX build, be sure that the configuration for flash organization matches on both.
> 
> ---

Reviewer 3: ```suggestion
{{< alert icon="eye" >}}
If MCUboot is built separately from the NuttX build, make sure that both have a matching flash partition table.
{{< /alert >}}
```

> Save and exit the menuconfig.

Reviewer 3: Not clear what this refers to. This line looks like part of a procedure, but from the previous content's formatting, it is not clear where the preceding steps of this procedure can be found.

> Build the MCUboot bootloader:
> 
> ```bash
> make -j bootloader
> ```
> 
> Now build the NuttX application:
> 
> ```bash
> make ESPTOOL_BINDIR=./ -s -j
> ```

Reviewer 3: This part looks like steps of the procedure. Please format these and the preceding steps so that they look like a procedure (not sure what the step numbers should be here, I randomly chose `4` to illustrate the formatting common for procedures):

````suggestion
4. Build the MCUboot bootloader:
  ```bash
  make -j bootloader
  ```
5. Now build the NuttX application:
  ```bash
  make ESPTOOL_BINDIR=./ -s -j
  ```
````

### Building the Reflashing application

> #### Example 1: Download the Target Reflashing Image

Reviewer 3: The same here, looks like this section has a procedure or a how-to, please format it as a procedure with distinct numbered steps and indented code blocks to show that they belong to this particular step.

> ESP Self-Reflasher component provides the example `boot_swap_download_example`, this can be used as start point for the Reflashing application.

Reviewer 3: Is _the Reflashing application_ a name of an actual application? Why is it capitalized?

```suggestion
The ESP Self-Reflasher component provides the example `boot_swap_download_example`. This can be used as a starting point for the Reflashing application.
```

```

Also configure the `Bootloader region address` (the Bootloader offset mentioned above) and the `Bootloader region size`.

Then configure the similar parameters on the `Example Configuration -> Application reflash configuration` menu, set the `Application upgrade url endpoint` (e.g. `http://192.168.0.100:8070/app_upd.bin`):

```
http://<HOST_IP>:<HOST_PORT>/app_upd.bin
```

Set the `Application region address` and `Application region size`. Note that if the `boot_swap_download_example` is flashed standalone directly as is, these must match the image Secondary Slot from the target MCUboot image organization as the example expects MCUboot to do the swapping to the Primary Slot later, otherwise, if the example is OTA upgraded which this tutorial walk through, these configs must match Primary Slot. It will depend on where the reflashing application is running from (Factory partition or OTA partition), as it should not overwrite itself on the process.

Configure the preferred network connection settings on the menu `Example Connection Configuration`, such as Wi-Fi SSID and Password.

Save and exit the menuconfig.

Build the example:

```bash
idf.py build
```

> #### Example 2: Embedded the Target Reflashing Image

Reviewer 3: Embedded > Embed?

The same here, looks like this section has a procedure or a how-to, please format it as a procedure with distinct numbered steps and indented code blocks to show that they belong to this particular step.

```bash
cd <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_embedded_example/
```

In this example, the MCUboot bootloader and target reflashing image needs to be merged in one binary as it goes embedded to the application.

```bash
esptool.py -c esp32 merge_bin --output <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_embedded_example/example_bin_dir/app_upd.merged.bin 0x0000 <NUTTX_DIR>/mcuboot-esp32.bin 0xF000 <NUTTX_DIR>/nuttx.bin
```

**Note** that in ESP32 case, the offsets on the merge step are shifted minus 0x1000 because of its bootloader offset 0x1000.

Configure the example:

```bash
idf.py menuconfig
```

Enter menu `Example Configuration` and set the directory where the merged binary is placed and its name. Set the `Destination address`, that for this example will be the Bootloader offset, and the `Destination region size`, that must be the image slot size from the target flash organization plus the bootloader region size.

Save and exit the menuconfig.

Build the example:

```bash
idf.py build
```

> ### Building and flashing the IDF application

Reviewer 3: The same here, looks like this section has a procedure or a how-to, please format it as a procedure with distinct numbered steps and indented code blocks to show that they belong to this particular step.

Reviewer 4: Should we standardize the ESP-IDF instead of IDF? I see many references to it just as IDF and not as ESP-IDF.

Save and exit the menuconfig.

Copy the Reflashing application binary that was build in the previous section to the directory where the HTTP server will run:

```bash
cp <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_<MODE>_example/build/boot_swap_<MODE>_example.bin <HTTP_SERVER_DIR>
```

Open a new bash and start the HTTP server for the OTA application. Here is a suggestion for creating the HTTP server using Python:

```bash
cd <HTTP_SERVER_DIR>
sudo python -m http.server -b <HOST_IP> <HOST_PORT>
# e.g. sudo python -m http.server -b 192.168.0.100 8070
```

---
If the `boot_swap_download_example` was used, copy the target reflashing images to the <HTTP_SERVER_DIR> or open another bash and start the HTTP server from where the reflashing images will be downloaded (watch for the <PORT> that cannot be the same):

```bash
cd <IDF_DIR>/components/esp-self-reflasher/examples/boot_swap_download_example/example_bin_dir/
sudo python -m http.server -b <HOST_IP> <HOST_PORT>
# e.g. sudo python -m http.server -b 192.168.0.100 8071
```

---

Finally build and flash the OTA application:

```bash
idf.py -p <SERIAL_PORT> flash monitor
# e.g. idf.py -p /dev/ttyUSB0 flash monitor
```

> ## Conclusion

Reviewer 3: Please add a paragraph about what has been done in the course of this article.

This will naturally lead to the existing paragraph that gives final suggestions before considering the migration.
